### DISCUSSION TITLE
How Do I Add A Tool?

### DISCUSSION THREAD

**Discussion by websmyths** [langflow-ai/langflow#27] [Q&A]
Is there a simple way to add a tool. E.G., I'd like to add the Wolfram wrapper to the list of Tools in the sidebar, and maybe some others. How would I accomplish that?

**Comment by ogabrielluiz** [comment]
Now there's a limited subset of langchain's functionality because we haven't tested all of it yet.

Some of them require specific python packages which pose their own challenge. 

Our focus will to test each one of them as quickly as possible and get at least the ones that are mainly API calls to langflow. 

Feel free to contribute if you manage to make some of them work.
The components available now are in this file:
https://github.com/logspace-ai/langflow/blob/472b0d657d8b2c086d4a0f47cb326dc43dad5404/langflow/backend/langflow_backend/utils/allowed_components.py

Adding a new one will give you access to it on the frontend which will allow you to test it and validate it.

**Reply by websmyths** [reply]
Thank you. Going to give it a shot. I've never contributed to a project, but I believe I know what to do. I'm confused about whether the project is using Poetry or Make now. I'll get back to you if I have questions.

**Reply by websmyths** [reply]
So it took me a minute or two to get the dev environment set up and working. At least I think I have it working correctly. The frontend fails to start up in the Docker container, but I can run it locally and that works. I added some more of the tools, but a couple won't show up in the UI. Wolfram Alpha and Wikipedia don't appear. But I was able to add Open Meteo API, News API, and TMDB API, and will test them. Any ideas about why Wolfram and Wikipedia won't appear and how I can debug it? I don't see any errors.

**Reply by websmyths** [reply]
Now that I think about it, they probably need packages installed. I'll look into that.

**Reply by ScripterSugar** [reply]
Maybe more general approach would help about this issue in the future? like creating UI to add custom tools/agents (I know this will be very complicated task to accomplish), that UI would cover up custom prompts (or even LLMs in the future for custom endpoints) so people can manage their own prompt templates much easier and seamlessly.

guess I should start search my own way to accomplish this.

**Reply by ogabrielluiz** [reply]
Our goal for this week is to add as many tools as possible and possibly some Memory modules too. 

They should work when activated as long as we add the dependencies. I'll see to adding an extra install option if possible to install langchain[all] which would install every package you might use in langchain. 

About the more general approach, we were thinking of adding a text area to add python code to it so you'd create the tools just like in LangChain.

**Reply by noahzev** [reply]
The FewShotTemplate includes a port for "Examples" which requires a "dict list" input, but there is no component to store a dict list to pass into that component.

**Reply by farrael004** [reply]
> we were thinking of adding a text area to add python code to it so you'd create the tools just like in LangChain.

Yes Please! I'd love to use my own API endpoints and integrate them inside of LanfFlow as tools. I think a great addition would be class wrapper nodes. They would be just empty implementations of the LangChain classes but you can put any code inside of them. They would still expect the same inputs and outputs.

**Reply by shuvracse03** [reply]
Is it possible to add a custom tool in langflow that is not in lagchain?

**Comment by cycloner2020** [comment]
This is an amazing tool! Is there integration to Pinecone as of today or is that a future module feature? Is there any sample code on how the flow method can be extended to other datasources such as Pinecone or Doc loaders?

**Comment by carlosrcoelho** [comment]
Hey!

Thanks for sharing your question here. I'm sorry for not getting back to you sooner.
Langflow has gone through some major changes since then. I recommend trying out the new version for the best experience. If you have any issues, please open a new one; we'll be here to help.
